link.go

```diff
@@ -304,69 +304,64 @@ func (l *link) setSettleModes(resp *frames.PerformAttach) error {
-				l.Source.Address, l.receiver.inFlight.len(), l.linkCredit, credits, drain, l.deliveryCount, len(l.Messages), l.countUnsettled(), l.receiver.maxCredit, l.ReceiverSettleMode.String())
-
-			// send a flow frame.
-			l.err = l.muxFlow(credits, drain)
-		}
-
-	// if receiver && half maxCredits have been processed, send more credits
-	case isReceiver && l.linkCredit+uint32(l.countUnsettled()) <= l.receiver.maxCredit/2:
-		debug(1, "FLOW Link Mux half: source: %s, inflight: %d, credit: %d, deliveryCount: %d, messages: %d, unsettled: %d, maxCredit : %d, settleMode: %s", l.Source.Address, l.receiver.inFlight.len(), l.linkCredit, l.deliveryCount, len(l.Messages), l.countUnsettled(), l.receiver.maxCredit, l.ReceiverSettleMode.String())
-
-		linkCredit := l.receiver.maxCredit - uint32(l.countUnsettled())
-		l.err = l.muxFlow(linkCredit, false)
+Loop:
+	for {
+		var outgoingTransfers chan frames.PerformTransfer
 
-		if l.err != nil {
-			return false, false
-		}
-		atomic.StoreUint32(&l.Paused, 0)
+		switch {
+		// enable outgoing transfers case if sender and credits are available
+		case isSender && l.linkCredit > 0:
+			debug(1, "Link Mux isSender: credit: %d, deliveryCount: %d, messages: %d, unsettled: %d", l.linkCredit, l.deliveryCount, len(l.Messages), l.countUnsettled())
+			outgoingTransfers = l.Transfers
 
-	case isReceiver && l.linkCredit == 0:
-		debug(1, "PAUSE Link Mux pause: inflight: %d, credit: %d, deliveryCount: %d, messages: %d, unsettled: %d, maxCredit : %d, settleMode: %s", l.receiver.inFlight.len(), l.linkCredit, l.deliveryCount, len(l.Messages), l.countUnsettled(), l.receiver.maxCredit, l.ReceiverSettleMode.String())
-		atomic.StoreUint32(&l.Paused, 1)
-	}
+		case isReceiver && l.receiver.manualCreditor != nil:
+			// TODO: fix the initial case to behave the same as the non-manual case
+			drain, credits := l.receiver.manualCreditor.FlowBits(l.linkCredit)
 
-	return true, false
-}
+			if drain || credits > 0 {
+				debug(1, "FLOW Link Mux (manual): source: %s, inflight: %d, credit: %d, creditsToAdd: %d, drain: %v, deliveryCount: %d, messages: %d, unsettled: %d, maxCredit : %d, settleMode: %s",
+					l.Source.Address, l.receiver.inFlight.len(), l.linkCredit, credits, drain, l.deliveryCount, len(l.Messages), l.countUnsettled(), l.receiver.maxCredit, l.ReceiverSettleMode.String())
 
-func (l *link) mux() {
-	defer l.muxDetach()
+				// send a flow frame.
+				l.err = l.muxFlow(credits, drain)
+				// TODO: consolidate error checks
+				if l.err != nil {
+					return
+				}
 
-Loop:
-	for {
-		var outgoingTransfers chan frames.PerformTransfer
+				// if muxFlow() decides to send a flow frame, it must "pump" l.RX to keep the
+				// session mux unblocked.  in the event that a frame is received at this point
+				// we must restart the loop so that we can reevaluate our link credit.
+				continue Loop
```

Comment:
For manual creditors do we need to loop again? There's no auto-adding of credits needed here, right?

Reply:
It's not adding credit, it's consuming it.  If `muxFlow()` receives a transfer frame while trying to send the flow frame we need to reevaluate the link credit to see if we should pause the link.

Reply:
So what you're going for is:

1. Hit the case at 325 
   ```go
   // ln 325
   case isReceiver && l.receiver.manualCreditor != nil
   ```
2. Some credits were consumed
3. Then hit the case at line 362?
   ```go
   ln 362
   case isReceiver && l.linkCredit == 0:
   ```

But if manual credits are enabled (and thus `receiver.manualCreditor` is always non-nil) we won't hit 362, right? Where will the pause bit get flipped in that case?

Reply:
Ugh you're correct.  In that case we will never pause the link.